fixes (#20938) by limiting embed previews to 2000

escherize opened the pull request:
fixes #20938

escherize committed:
fixes (#20938) by limiting embed previews to 2000

- (so they load faster)

escherize committed:
adds test for (#20938)

escherize committed:
add comment to the test for (#20938)

camsaul commented on test/metabase/api/preview_embed_test.clj:
```diff
@@ -141,6 +141,20 @@
+      (embed-test/with-embedding-enabled-and-new-secret-key
+        (let [;; this is the same sample question as setup in these steps to reproduce:
+              ;; https://github.com/metabase/metabase/issues/20938
+              sample-db-orders-question {:database 1
+                                         :query {:source-table 2}
+                                         :type "query"}]
```
You can't assume that this DB exists yet or that it's ID is going to be `1`... what ID it gets depends on the order tests are ran. As proof, try

```sh
clojure -X:dev:test :only metabase.api.preview-embed-test/query-max-results-constraint-test
```

I'm not sure which Table you expect to get for ID `2`, but you want to do something like

`(mt/dataset sample-dataset ...)` and `(mt/mbql-query <whatever-table-name>)`. Grep the codebase for `mt/dataset` and you should find tons of examples

camsaul replied:
Of course this only works if the table in question has > 2000 rows

camsaul commented on test/metabase/api/preview_embed_test.clj:
```diff
@@ -141,6 +141,20 @@
+                                         :query {:source-table 2}
+                                         :type "query"}]
+          (embed-test/with-temp-card [card {:dataset_query sample-db-orders-question}]
+            (is (>= 2000 (-> (mt/user-http-request :crowberto :get 202 (card-query-url card))
```
I think this would be a bit more readable if you wrote it at `(<= row-count 2000)`. Otherwise it's too much of a [Yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions). You would read `(>= 2000 x)` as `2000 >= x` which is a little _weird_.  `x <= 2000` is more natural to read. I know equality tests we do `(is (= expected actual))` but this is a normal predicate test expression so I think the Yoda condition rule applies here.

I'm only brining this up because I did a double take and spent a couple minutes reading this to make sure it was right when I saw it. YMMV

escherize replied:
Yeah, as you could tell I was matching the `(is (= expected actual))` pattern. This has been changed so as to not use `<=` or `>=`.

camsaul commented on test/metabase/api/preview_embed_test.clj:
```diff
@@ -141,6 +141,20 @@
+                                         :type "query"}]
+          (embed-test/with-temp-card [card {:dataset_query sample-db-orders-question}]
+            (is (>= 2000 (-> (mt/user-http-request :crowberto :get 202 (card-query-url card))
+                             :data
+                             :rows
```
Instead of `(-> ... :data :rows)` you should use `mt/rows` which does the same thing but throws a useful exception with the query error if the query failed. In that case, there is no `[:data :rows]` so `(count nil)` will just return `0` and your test failure will be 

```
(not (>= 2000 0))
```

which is not easy to debug. `mt/rows` will give you the actual query error in the failure

camsaul replied:
```suggestion
                             mt/rows
```

camsaul reviewed (changes requested):
1. Hardcoding Database/Table/Field IDs is illegal according to the laws of writing Metabase backend tests 
2. The Yoda condition thing is my personal opinion and if you don't agree in this case you don't need to change it
3. `(mt/rows x)` is better than `(:rows (:data x))` for debugging failing tests
4. No magic numbers please

camsaul commented on src/metabase/api/preview_embed.clj:
```diff
@@ -37,6 +37,7 @@
       :card-id          card-id
       :token-params     (eu/get-in-unsigned-token-or-throw unsigned-token [:params])
       :embedding-params (eu/get-in-unsigned-token-or-throw unsigned-token [:_embedding_params])
+      :constraints      {:max-results 2000}
```
One more thing: this is definitely a [magic number](https://en.wikipedia.org/wiki/Magic_number_(programming)) and it should go in a `def` somewhere rather than just living there right in the function call. As a bonus that would also let you use `with-redefs` to change it to make things easier to test, for example you can redefine it to 5 and make sure a table only returns 5 rows. That's easier (and a lot faster) than making sure a table (a) has more than 2000 rows and (b) returns the first 2000 rows. 

Even if you don't change the test to use `with-redefs`, the same magic number rule applies here. You should refer to the variable instead of hard-coding 2000. Why? Think about what happens if we change this number to something like 500 in the future. The test would start failing, and you'd have to go update the number there as well. Not a big deal here since it's only two places to update, but if we had say 30 places using the number we'd have 30 places to update and so forth. Better to avoid magic numbers for that reason (and also because it makes code more readable).

escherize committed:
Update test/metabase/api/preview_embed_test.clj

Co-authored-by: Cam Saul <1455846+camsaul@users.noreply.github.com>

escherize committed:
respond to code review

- yoda condition get rid of by using = and not=
- use mt/rows !
- don't hardcode table ids
- remove magic 2000 number (this helped with the test)

escherize committed:
include metabase.api.preview-embed ns in test

escherize committed:
fix ns ordering in test

escherize committed:
Merge branch 'master' into limit-preview-embedding-result-count

camsaul reviewed (approved):
Awesome this LGTM
